docs/manual/resource-leak-checker.tex

```diff
@@ -588,6 +597,234 @@
+
+\begin{verbatim}
+Example.java:31: error: [required.method.not.called] @MustCall method close may not have been invoked on new Socket(myHost, myPort) or any of its aliases.
+            arr[0] = new Socket(myHost, myPort);
+                     ^
+  The type of object is: java.net.Socket.
+  Reason for going out of scope: regular method exit
+1 error
+\end{verbatim}
+
+We need a way for the array to take on this obligation for the resource. This functionality is jointly provided by the \<MustCallOnElements> and \<CalledMethodsOnElements> subcheckers.
+
+\subsectionAndLabel{Declaration and assignment of resource-holding arrays}{resource-leak-arrays-declaration}
+Annotating the array with the \<@OwningArray> declaration annotation allows it to take on such obligations by using a pattern-matched assignment syntax:
+
+\begin{verbatim}
+@OwningArray Socket[] s = new Socket[n];
+// 1. @MustCallOnElements({}) @CalledMethodsOnElements({})
+for (int i = 0; i < s.length; i++) {
+  try {
+    s[i] = new Socket(myHost, myPort);
+  } catch (Exception e) {
+    // do nothing
+  }
+}
+// 2. @MustCallOnElements({"close"}) @CalledMethodsOnElements({})
+\end{verbatim}
+The loop must iterate from 0 to \<s.length> or \<n>, where \<n> is the effectively final variable used for the initialization of the array size. The update must be either a pre-increment or post-increment.  There must be exactly one statement in the body (which may be a try-catch construct) that contains the assignment. The right-hand side of the assignment must be the immediate construction of the object (not simply any expression evaluating to one). If there is a try-catch construct, the catch and finally blocks may not throw any exceptions, break out of the loop, call any methods or return for successful pattern-matching. In order to assign the elements of an \<@OwningArray>, the array must have no open calling obligations on its elements.
+
+\noindent Any other assignment to the elements of an \<@OwningArray>, outside of such a loop, is prohibited.
+
+An \<@OwningArray> reference may be reassigned after an initial assignment, but only to allocate a new array (and of course only if the array has no open must-call-on-elements obligations). Any other assignment with the left-hand side being an \<@OwningArray> reference is prohibited. In particular is an aliasing of two \<@OwningArray> references prohibited:
+
+\begin{verbatim}
+@OwningArray Socket[] s1 = new Socket[n];
+@OwningArray Socket[] s2 = new Socket[n];
+s2 = s1; // prohibited assignment
+\end{verbatim}
+
+In general is aliasing of any kind for \<@OwningArray> variables not permitted. In particular, a non-\<@OwningArray> variable may also not be assigned to an \<@OwningArray> variable. This implies that an \<@OwningArray> variable is always the only reference to the underlying array within a procedure. The exception occurs when an \<@OwningArray> referenc is passed as an argument. More on that later.
```

Comment:
Why is it prohibited to assign an `@OwningArray` variable to another variable that is non-`@OwningArray`?

Reply:
Nevermind, I see now that the issue is writes through the non-`@OwningArray` alias, which could e.g. null out an array index without closing the contained element first